teps/0044-composing-tasks-with-tasks.md

```diff
@@ -0,0 +1,137 @@
+by [PipelineResources](https://github.com/tektoncd/pipeline/blob/master/docs/resources.md).
+[We did not bring these types to beta](https://github.com/tektoncd/pipeline/blob/master/docs/resources.md#why-arent-pipelineresources-in-beta).
+Issues that don't let PipelineResources (as is) solve these problems are:
+* PipelineResource "outputs" do not run when the steps of the Task fail (or if
+  a previous PipelienResource "output" fails) (see [unit test use case](#use-cases-optional))
+* PipelineResources do not use Tasks (which is something we could change), meaning
+  you cannot use them to compose Tasks together, you need to build PipelineResources
+  * Adding new PipelineResources currently needs to be done in the Tekton Pipelines controller,
+    tho there have been several attempts to propose ways to make it extensible:
+    * [PipelineResources 2 Uber Design Doc](https://docs.google.com/document/d/1euQ_gDTe_dQcVeX4oypODGIQCAkUaMYQH5h7SaeFs44/edit)
+    * [Tekton Pipeline Resource Extensibility](https://docs.google.com/document/d/1rcMG1cIFhhixMSmrT734MBxvH3Ghre9-4S2wbzodPiU/edit)
+  * All of this is even though PipelineResources are _very_ similar to Tasks in that they are
+    collections of steps (which [get injected before and after a Task's steps](https://github.com/tektoncd/pipeline/issues/1838)
+* Tasks have to declare in advance what PipelineResources they need; you can't decide to use
+  a PipelineResource with a Task after it has been written (e.g. in a TaskRun or a Pipeline)
+  and you can't mix types of PipelineResource (e.g. if a Task declares it needs a
+  [git PipelineResources](https://github.com/tektoncd/pipeline/blob/master/docs/resources.md#git-resource)
+  it can't use a different kind of PipelineResource to provide those files (tho you can avoid
+  doing the clone multiple times using [from](https://github.com/tektoncd/pipeline/blob/master/docs/pipelines.md#using-the-from-parameter))
+
+### Goals
+
+- Make it possible to combine Tasks together so that you can run multiple
+  Tasks as "one unit" (see [Requirements](#requirements):
+  - At runtime in a TaskRun
+  - At Pipeline authoring time (which would ultimately be realized through TaskRuns)
+- Add some of [the features we don't have without PipelineResources](https://docs.google.com/document/d/1KpVyWi-etX00J3hIz_9HlbaNNEyuzP6S986Wjhl3ZnA/edit#)
+  to Tekton Pipelines (without requiring use of PipelineResources)
+
+### Non-Goals
+
+- Composing Tasks within Tasks at [Task authoring time](https://github.com/tektoncd/community/blob/master/design-principles.md#reusability).
+  We could decide to include this if we have some use cases that need it; for now avoiding
+  this allows us to avoid many layers of nesting (i.e. Task1 uses Task2 uses Task3, etc.)
+  or even worse, recursion (Task 1 uses Task 2 uses Task 1...)
+  - One way to support this later could be via 
+- Supporting this composition at runtime in a PipelineRun (not quite sure what that
```

Comment:
I guess it depends on how it is implemented - it may be possible to use `pipelineSpec` - but I agree this should not be a direct goal.